<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 16 DEG analysis | 2022년 한국생명공학연구원 연구데이터 분석과정 R</title>
<meta name="author" content="합성생물학전문연구소 김하성">
<meta name="description" content="16.1 High-throughput genomic data analysis   16.2 DEG analysis with bioconductor Differentially Expressed Gene 분석은 전통적 two-channel microarray나 RNA-Seq 데이터를 활용한 분석입니다. Genome reference에 fastq 파일의...">
<meta name="generator" content="bookdown 0.26 with bs4_book()">
<meta property="og:title" content="Chapter 16 DEG analysis | 2022년 한국생명공학연구원 연구데이터 분석과정 R">
<meta property="og:type" content="book">
<meta property="og:description" content="16.1 High-throughput genomic data analysis   16.2 DEG analysis with bioconductor Differentially Expressed Gene 분석은 전통적 two-channel microarray나 RNA-Seq 데이터를 활용한 분석입니다. Genome reference에 fastq 파일의...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 16 DEG analysis | 2022년 한국생명공학연구원 연구데이터 분석과정 R">
<meta name="twitter:description" content="16.1 High-throughput genomic data analysis   16.2 DEG analysis with bioconductor Differentially Expressed Gene 분석은 전통적 two-channel microarray나 RNA-Seq 데이터를 활용한 분석입니다. Genome reference에 fastq 파일의...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/kePrint-0.0.1/kePrint.js"></script><link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">2022년 한국생명공학연구원 연구데이터 분석과정 R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="rrstudio-basics.html"><span class="header-section-number">2</span> R/Rstudio basics</a></li>
<li><a class="" href="rmarkdown.html"><span class="header-section-number">3</span> Rmarkdown</a></li>
<li><a class="" href="r-programming.html"><span class="header-section-number">4</span> R programming</a></li>
<li><a class="" href="data-transformation.html"><span class="header-section-number">5</span> Data transformation</a></li>
<li><a class="" href="r-basic-graphics.html"><span class="header-section-number">6</span> R basic graphics</a></li>
<li><a class="" href="tidyverse.html"><span class="header-section-number">7</span> tidyverse</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">8</span> Data visualization</a></li>
<li><a class="" href="ggplot2-examples.html"><span class="header-section-number">9</span> ggplot2 examples</a></li>
<li><a class="" href="bioconductor.html"><span class="header-section-number">10</span> Bioconductor</a></li>
<li><a class="" href="biostrings.html"><span class="header-section-number">11</span> Biostrings</a></li>
<li><a class="" href="tools-for-sequences.html"><span class="header-section-number">12</span> Tools for sequences</a></li>
<li><a class="" href="tools-for-genome-analysis.html"><span class="header-section-number">13</span> Tools for genome analysis</a></li>
<li><a class="" href="high-throughput-data.html"><span class="header-section-number">14</span> High-throughput data</a></li>
<li><a class="" href="inference-and-test.html"><span class="header-section-number">15</span> Inference and test</a></li>
<li><a class="active" href="deg-analysis.html"><span class="header-section-number">16</span> DEG analysis</a></li>
<li><a class="" href="lecture-note.html"><span class="header-section-number">17</span> Lecture Note</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="deg-analysis" class="section level1" number="16">
<h1>
<span class="header-section-number">16</span> DEG analysis<a class="anchor" aria-label="anchor" href="#deg-analysis"><i class="fas fa-link"></i></a>
</h1>
<div id="high-throughput-genomic-data-analysis" class="section level2" number="16.1">
<h2>
<span class="header-section-number">16.1</span> High-throughput genomic data analysis<a class="anchor" aria-label="anchor" href="#high-throughput-genomic-data-analysis"><i class="fas fa-link"></i></a>
</h2>
<div class="inline-figure"><img src="images/12/htanalysis.png" width="600"></div>
</div>
<div id="deg-analysis-with-bioconductor" class="section level2" number="16.2">
<h2>
<span class="header-section-number">16.2</span> DEG analysis with bioconductor<a class="anchor" aria-label="anchor" href="#deg-analysis-with-bioconductor"><i class="fas fa-link"></i></a>
</h2>
<p>Differentially Expressed Gene 분석은 전통적 two-channel microarray나 RNA-Seq 데이터를 활용한 분석입니다. Genome reference에 fastq 파일의 read들을 맵핑하고 맵핑된 read들을 카운팅하여 해당 유전자의 발현을 정량화하고 이를 기준이 되는 발현값과 비교하여 질병이나 조건에 따라 다른 발현값을 갖는 유전자를 찾는 방법입니다.</p>
<p>Reference 서열 정보와 발현된 mRNA 서열을 분석한 fastq 파일이 필요하며 BWA, Bowtie, tophat 등의 linux 스크립트 기반 software들이 있으며 Bioconductor에서도 다양한 분석 툴이 소개되고 있습니다. 참고로 진핵세포의 경우 splicing 등을 고려한 mapping 기술이 필요합니다.</p>
</div>
<div id="creating-a-reference-genome" class="section level2" number="16.3">
<h2>
<span class="header-section-number">16.3</span> Creating a reference genome<a class="anchor" aria-label="anchor" href="#creating-a-reference-genome"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb326"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">genbankr</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span>url<span class="op">=</span><span class="st">"https://github.com/greendaygh/kribbr2022/raw/main/ecoli-mg1655.gb"</span>, destfile<span class="op">=</span><span class="st">"examples/ecoli-mg1655.gb"</span><span class="op">)</span>

<span class="va">ecoli</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/genbankr/man/readGenBank.html">readGenBank</a></span><span class="op">(</span><span class="st">"examples/ecoli-mg1655.gb"</span><span class="op">)</span>
<span class="va">ecoliseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html">getSeq</a></span><span class="op">(</span><span class="va">ecoli</span><span class="op">)</span>


<span class="va">ecoliseqsub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/XVector/man/XVector-class.html">subseq</a></span><span class="op">(</span><span class="va">ecoliseq</span>, <span class="fl">1</span>, <span class="fl">100000</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">ecoliseqsub</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"K-12"</span>
<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">writeXStringSet</a></span><span class="op">(</span><span class="va">ecoliseqsub</span>, <span class="st">"examples/ecolisub.fasta"</span><span class="op">)</span></code></pre></div>
</div>
<div id="read-alignment-and-mapping" class="section level2" number="16.4">
<h2>
<span class="header-section-number">16.4</span> Read Alignment and Mapping<a class="anchor" aria-label="anchor" href="#read-alignment-and-mapping"><i class="fas fa-link"></i></a>
</h2>
<p>Rsubread 패키지를 활용해서 align을 수행합니다. align 전에 reference genome의 index를 생성해야 하며 지놈이 클 경우 오랜 시간이 소요될 수 있으니 주의가 필요합니다.</p>
<div class="sourceCode" id="cb327"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"Rsubread"</span><span class="op">)</span></code></pre></div>
<p>index 생성은 Rsubread 패키지의 buildindex 함수를 사용하며 memory 옵션을 이용해서 데이터 사이즈에 따라 할당되는 메모리를 조절할 수 있습니다. 기본은 8GB로 memory = 8000 입니다. indexSplit 옵션으로 크로모좀별로 인덱스를 분리해서 메모리 효율을 증가시킬 수 있습니다.</p>
<div class="sourceCode" id="cb328"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/Rsubread">Rsubread</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/Rsubread/man/buildindex.html">buildindex</a></span><span class="op">(</span>basename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"examples"</span>, <span class="st">"ecoliexample"</span><span class="op">)</span>, 
           reference <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"examples"</span>, <span class="st">"ecolisub.fasta"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="rna-seq-alignment-mapping" class="section level2" number="16.5">
<h2>
<span class="header-section-number">16.5</span> RNA-Seq alignment (Mapping)<a class="anchor" aria-label="anchor" href="#rna-seq-alignment-mapping"><i class="fas fa-link"></i></a>
</h2>
<p>Rsubread 패키지를 활용해서 mapping을 수행합니다. align 함수를 사용하며 splicing 여부에 따라 옵션이 조금씩 다를 수 있습니다.</p>
<div class="sourceCode" id="cb329"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/Rsubread">Rsubread</a></span><span class="op">)</span>

<span class="va">alignstat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsubread/man/align.html">align</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"examples"</span>, <span class="st">"ecoliexample"</span><span class="op">)</span>
                   , readfile1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"examples"</span>, <span class="st">"filtered_SRR11549076_1.fastq_R1.fastq.gz"</span><span class="op">)</span>
                   , output_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"examples"</span>, <span class="st">"ecoliexample.BAM"</span><span class="op">)</span>
                   , nthreads <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>


<span class="co">#alignstat</span></code></pre></div>
</div>
<div id="sorting" class="section level2" number="16.6">
<h2>
<span class="header-section-number">16.6</span> sorting<a class="anchor" aria-label="anchor" href="#sorting"><i class="fas fa-link"></i></a>
</h2>
<p>SAM 파일은 Sequence alignment data를 담고 있는 텍스트 파일(.txt)로 각 내용은 탭(tab)으로 분리되어 alignment, mapping 정보를 담고 있습니다. BAM 파일은 SAM 파일의 binary 버전으로 동일한 정보를 담고 있으며 이들 파일을 다루기 위해서는 SAMtools 소프트웨어가 필요합니다. R에서는 SAMtools의 R 버전인 Rsamtools 패키지를 활용할 수 있습니다.</p>
<div class="sourceCode" id="cb330"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools">Rsamtools</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html">sortBam</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"examples"</span>, <span class="st">"ecoliexample.BAM"</span><span class="op">)</span>
        , destination <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"examples"</span>, <span class="st">"sorted_ecoliexample.BAM"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html">indexBam</a></span><span class="op">(</span>files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"examples"</span>, <span class="st">"sorted_ecoliexample.BAM.bam"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="visualization-1" class="section level2" number="16.7">
<h2>
<span class="header-section-number">16.7</span> visualization<a class="anchor" aria-label="anchor" href="#visualization-1"><i class="fas fa-link"></i></a>
</h2>
<p><a href="https://software.broadinstitute.org/software/igv/">IGV</a>를 활용하여 mapping 파일 가시화가 가능합니다.</p>
</div>
<div id="counting-in-gene-models" class="section level2" number="16.8">
<h2>
<span class="header-section-number">16.8</span> Counting in gene models<a class="anchor" aria-label="anchor" href="#counting-in-gene-models"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb331"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicAlignments">GenomicAlignments</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span>

<span class="va">ecolicds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html">cds</a></span><span class="op">(</span><span class="va">ecoli</span><span class="op">)</span>
<span class="va">ecolicds_sub</span> <span class="op">&lt;-</span> <span class="va">ecolicds</span> <span class="op"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/tidyverse-reexports.html">filter</a></span><span class="op">(</span><span class="va">end</span> <span class="op">&lt;</span> <span class="fl">100000</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqlengths</a></span><span class="op">(</span><span class="va">ecolicds_sub</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">100000</span>


<span class="va">mybam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html">BamFile</a></span><span class="op">(</span><span class="st">"examples/sorted_ecoliexample.BAM.bam"</span>, yieldSize <span class="op">=</span> <span class="fl">100000</span><span class="op">)</span>
<span class="va">myresult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicAlignments/man/summarizeOverlaps-methods.html">summarizeOverlaps</a></span><span class="op">(</span><span class="va">ecolicds_sub</span>, <span class="va">mybam</span>, ignore.strand <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">myresult</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">assay</a></span><span class="op">(</span><span class="va">myresult</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span><span class="op">(</span><span class="va">myresult</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html">colData</a></span><span class="op">(</span><span class="va">myresult</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata</a></span><span class="op">(</span><span class="va">myresult</span><span class="op">)</span></code></pre></div>
<hr>
<p><a rel="license" href="http://creativecommons.org/licenses/by-nc-nd/4.0/"><img alt="크리에이티브 커먼즈 라이선스" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-nd/4.0/88x31.png"></a><br>이 저작물은 <a rel="license" href="http://creativecommons.org/licenses/by-nc-nd/4.0/">크리에이티브 커먼즈 저작자표시-비영리-변경금지 4.0 국제 라이선스</a>에 따라 이용할 수 있습니다.</p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="inference-and-test.html"><span class="header-section-number">15</span> Inference and test</a></div>
<div class="next"><a href="lecture-note.html"><span class="header-section-number">17</span> Lecture Note</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#deg-analysis"><span class="header-section-number">16</span> DEG analysis</a></li>
<li><a class="nav-link" href="#high-throughput-genomic-data-analysis"><span class="header-section-number">16.1</span> High-throughput genomic data analysis</a></li>
<li><a class="nav-link" href="#deg-analysis-with-bioconductor"><span class="header-section-number">16.2</span> DEG analysis with bioconductor</a></li>
<li><a class="nav-link" href="#creating-a-reference-genome"><span class="header-section-number">16.3</span> Creating a reference genome</a></li>
<li><a class="nav-link" href="#read-alignment-and-mapping"><span class="header-section-number">16.4</span> Read Alignment and Mapping</a></li>
<li><a class="nav-link" href="#rna-seq-alignment-mapping"><span class="header-section-number">16.5</span> RNA-Seq alignment (Mapping)</a></li>
<li><a class="nav-link" href="#sorting"><span class="header-section-number">16.6</span> sorting</a></li>
<li><a class="nav-link" href="#visualization-1"><span class="header-section-number">16.7</span> visualization</a></li>
<li><a class="nav-link" href="#counting-in-gene-models"><span class="header-section-number">16.8</span> Counting in gene models</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>2022년 한국생명공학연구원 연구데이터 분석과정 R</strong>" was written by 합성생물학전문연구소 김하성. It was last built on 2022-10-06.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
