<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 12 Tools for sequences | 2022년 한국생명공학연구원 연구데이터 분석과정 R</title>
<meta name="author" content="합성생물학전문연구소 김하성">
<meta name="description" content="12.1 Sequences from NCBI 전세계 연구자들이 서열 데이터를 분석하는데 가장 많이 이용하는 사이트 중 하나가 NCBI 이며 따라서 NCBI에서는 연구자들이 데이터베이스에 접근하기위한 편리한 방법을 제공하고 있고 그 중 하나가 Entrez 입니다. R에서도 Entrez 기능을 도입한 package들이 제공되고 있으며 그 중 하나가...">
<meta name="generator" content="bookdown 0.26 with bs4_book()">
<meta property="og:title" content="Chapter 12 Tools for sequences | 2022년 한국생명공학연구원 연구데이터 분석과정 R">
<meta property="og:type" content="book">
<meta property="og:description" content="12.1 Sequences from NCBI 전세계 연구자들이 서열 데이터를 분석하는데 가장 많이 이용하는 사이트 중 하나가 NCBI 이며 따라서 NCBI에서는 연구자들이 데이터베이스에 접근하기위한 편리한 방법을 제공하고 있고 그 중 하나가 Entrez 입니다. R에서도 Entrez 기능을 도입한 package들이 제공되고 있으며 그 중 하나가...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 12 Tools for sequences | 2022년 한국생명공학연구원 연구데이터 분석과정 R">
<meta name="twitter:description" content="12.1 Sequences from NCBI 전세계 연구자들이 서열 데이터를 분석하는데 가장 많이 이용하는 사이트 중 하나가 NCBI 이며 따라서 NCBI에서는 연구자들이 데이터베이스에 접근하기위한 편리한 방법을 제공하고 있고 그 중 하나가 Entrez 입니다. R에서도 Entrez 기능을 도입한 package들이 제공되고 있으며 그 중 하나가...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/kePrint-0.0.1/kePrint.js"></script><link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">2022년 한국생명공학연구원 연구데이터 분석과정 R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Introduction</a></li>
<li><a class="" href="rrstudio-basics.html"><span class="header-section-number">2</span> R/Rstudio basics</a></li>
<li><a class="" href="rmarkdown.html"><span class="header-section-number">3</span> Rmarkdown</a></li>
<li><a class="" href="r-programming.html"><span class="header-section-number">4</span> R programming</a></li>
<li><a class="" href="data-transformation.html"><span class="header-section-number">5</span> Data transformation</a></li>
<li><a class="" href="r-basic-graphics.html"><span class="header-section-number">6</span> R basic graphics</a></li>
<li><a class="" href="tidyverse.html"><span class="header-section-number">7</span> tidyverse</a></li>
<li><a class="" href="data-visualization.html"><span class="header-section-number">8</span> Data visualization</a></li>
<li><a class="" href="ggplot2-examples.html"><span class="header-section-number">9</span> ggplot2 examples</a></li>
<li><a class="" href="bioconductor.html"><span class="header-section-number">10</span> Bioconductor</a></li>
<li><a class="" href="biostrings.html"><span class="header-section-number">11</span> Biostrings</a></li>
<li><a class="active" href="tools-for-sequences.html"><span class="header-section-number">12</span> Tools for sequences</a></li>
<li><a class="" href="tools-for-genome-analysis.html"><span class="header-section-number">13</span> Tools for genome analysis</a></li>
<li><a class="" href="high-throughput-data.html"><span class="header-section-number">14</span> High-throughput data</a></li>
<li><a class="" href="inference-and-test.html"><span class="header-section-number">15</span> Inference and test</a></li>
<li><a class="" href="deg-analysis.html"><span class="header-section-number">16</span> DEG analysis</a></li>
<li><a class="" href="deg-excercise.html"><span class="header-section-number">17</span> DEG Excercise</a></li>
<li><a class="" href="lecture-note.html"><span class="header-section-number">18</span> Lecture Note</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="tools-for-sequences" class="section level1" number="12">
<h1>
<span class="header-section-number">12</span> Tools for sequences<a class="anchor" aria-label="anchor" href="#tools-for-sequences"><i class="fas fa-link"></i></a>
</h1>
<div id="sequences-from-ncbi" class="section level2" number="12.1">
<h2>
<span class="header-section-number">12.1</span> Sequences from NCBI<a class="anchor" aria-label="anchor" href="#sequences-from-ncbi"><i class="fas fa-link"></i></a>
</h2>
<p>전세계 연구자들이 서열 데이터를 분석하는데 가장 많이 이용하는 사이트 중 하나가 NCBI 이며 따라서 NCBI에서는 연구자들이 데이터베이스에 접근하기위한 편리한 방법을 제공하고 있고 그 중 하나가 Entrez 입니다.</p>
<p>R에서도 Entrez 기능을 도입한 package들이 제공되고 있으며 그 중 하나가 <code>rentrez</code> 입니다. <a href="https://www.ncbi.nlm.nih.gov/books/NBK25500/" class="uri">https://www.ncbi.nlm.nih.gov/books/NBK25500/</a> 이 곳의 Downloading Full Records 를 참고하시면 좋습니다. Entrez는 대략적으로 다음 9개의 유틸리티를 제공합니다.</p>
<blockquote>
<p>EInfo (database statistics)<br>
ESearch (text searches)<br>
EPost (UID uploads)<br>
ESummary (document summary downloads)<br>
EFetch (data record downloads)<br>
ELink (Entrez links)<br>
EGQuery (global query)<br>
ESpell (spelling suggestions)<br>
ECitMatch (batch citation searching in PubMed)</p>
</blockquote>
<p>이 중 <code>ESerach</code>, <code>EPost</code>, <code>ESummary</code>, <code>EFetch</code> 등이 많이 사용하는 유틸이며 정보를 다운로드 받을 경우는 <code>EFetch</code> 를 주로 사용하게 됩니다. rentrez 는 위와 같은 NCBI Eutils API를 활용하여 R 환경에서 탐색이나 다운로드 등 NCBI 데이터베이스와 상호작용이 용이하도록 만들어 놓은 tool 입니다. <a href="https://cran.r-project.org/web/packages/rentrez/vignettes/rentrez_tutorial.html">rentrez landing page</a> <code>entrez_dbs</code>명령은 NCBI에서 제공하는 데이터베이스의 리스트를 볼 수 있으며 특정 DB에 대한 설명은 <code>entrez_db_summary</code>를 사용하면 되겠습니다. <code>entrez_search</code>는 각종 키워드를 사용한 검색 기능을 제공합니다.</p>
<div class="sourceCode" id="cb250"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/rentrez/">rentrez</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings">Biostrings</a></span><span class="op">)</span>

<span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_dbs.html">entrez_dbs</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_db_summary.html">entrez_db_summary</a></span><span class="op">(</span><span class="st">"nuccore"</span><span class="op">)</span>

<span class="va">covid_paper</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_search.html">entrez_search</a></span><span class="op">(</span>db<span class="op">=</span><span class="st">"pubmed"</span>, term<span class="op">=</span><span class="st">"covid19"</span><span class="op">)</span>
<span class="va">covid_paper</span><span class="op">$</span><span class="va">ids</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">covid_paper</span><span class="op">)</span>
<span class="va">covid_paper</span><span class="op">$</span><span class="va">ids</span>


<span class="va">covid_link</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_link.html">entrez_link</a></span><span class="op">(</span>db<span class="op">=</span><span class="st">"all"</span>, id<span class="op">=</span><span class="va">covid_paper</span><span class="op">$</span><span class="va">ids</span>, dbfrom<span class="op">=</span><span class="st">"pubmed"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">covid_link</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">covid_link</span><span class="op">$</span><span class="va">links</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">head</a></span><span class="op">(</span><span class="va">covid_link</span><span class="op">$</span><span class="va">links</span><span class="op">$</span><span class="va">pubmed_pubmed</span><span class="op">)</span></code></pre></div>
<p><code>entrez_search</code>에서 검색어를 입력하는 방식은 <a href="https://cran.r-project.org/web/packages/rentrez/vignettes/rentrez_tutorial.html#building-search-terms">이곳</a>을 참고하세요. 검색으로 찾아진 특정 오브젝트(객체)에 대한 내용은 <code>entrez_summary</code> 함수를 사용하여 조회할 수 있으며 <code>extract_from_esummary</code>로 조회된 아이템들에 대한 정보를 추출할 수 있습니다. 특정 id에 대한 서열 등 다양한 타입의 데이터를 실제로 다운로드 받는 기능은 <code>entrez_fetch</code> 함수가 제공하고 있습니다. <code>entrez_fetch</code> 함수의 <code>rettype</code> 옵션에서 지원하는 데이터 타입을 다운로드 받을 수 있으며 rettype (return type)의 자세한 정보는 <a href="https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/">Eutils table</a> 또는 <a href="https://www.ncbi.nlm.nih.gov/books/NBK25499/">NCBI Eutils</a> 페이지를 참고하시기 바랍니다.</p>
<div class="sourceCode" id="cb251"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># popset database is a collection of related DNA sequences derived from population</span>
<span class="va">katipo_search</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_search.html">entrez_search</a></span><span class="op">(</span>db<span class="op">=</span><span class="st">"popset"</span>, term<span class="op">=</span><span class="st">"Latrodectus katipo[Organism]"</span><span class="op">)</span>
<span class="va">katipo_search</span><span class="op">$</span><span class="va">ids</span>

<span class="va">katipo_summs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_summary.html">entrez_summary</a></span><span class="op">(</span>db<span class="op">=</span><span class="st">"popset"</span>, id<span class="op">=</span><span class="va">katipo_search</span><span class="op">$</span><span class="va">ids</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">katipo_summs</span><span class="op">)</span>
<span class="va">katipo_summs</span><span class="op">$</span><span class="va">`41350664`</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">katipo_summs</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/methods.html">methods</a></span><span class="op">(</span>class<span class="op">=</span><span class="st">"esummary_list"</span><span class="op">)</span>

<span class="va">titles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/extract_from_esummary.html">extract_from_esummary</a></span><span class="op">(</span><span class="va">katipo_summs</span>, <span class="st">"title"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/unname.html">unname</a></span><span class="op">(</span><span class="va">titles</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">katipo_summs</span><span class="op">)</span>
<span class="va">katipo_summs</span><span class="op">$</span><span class="va">`1790798044`</span><span class="op">$</span><span class="va">gi</span>


<span class="va">COI_ids</span> <span class="op">&lt;-</span> <span class="va">katipo_search</span><span class="op">$</span><span class="va">ids</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">6</span><span class="op">)</span><span class="op">]</span>
<span class="va">trnL_ids</span> <span class="op">&lt;-</span> <span class="va">katipo_search</span><span class="op">$</span><span class="va">ids</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>
<span class="va">COI</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_fetch.html">entrez_fetch</a></span><span class="op">(</span>db<span class="op">=</span><span class="st">"popset"</span>, id<span class="op">=</span><span class="va">COI_ids</span>, rettype<span class="op">=</span><span class="st">"fasta"</span><span class="op">)</span>
<span class="va">trnL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_fetch.html">entrez_fetch</a></span><span class="op">(</span>db<span class="op">=</span><span class="st">"popset"</span>, id<span class="op">=</span><span class="va">trnL_ids</span>, rettype<span class="op">=</span><span class="st">"fasta"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/write.html">write</a></span><span class="op">(</span><span class="va">COI</span>, <span class="st">"COI.fasta"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/write.html">write</a></span><span class="op">(</span><span class="va">trnL</span>, <span class="st">"trnl.fasta"</span><span class="op">)</span>

<span class="co">#library(Biostrings)</span>
<span class="va">coi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="st">"COI.fasta"</span><span class="op">)</span>
<span class="va">trnl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="st">"trnl.fasta"</span><span class="op">)</span></code></pre></div>
<div class="rmdnote">
<p><strong>Exercises </strong></p>
<p>뎅기바이러스 서열 4종에 대한 NCBI의 accession 번호가 다음과 같음 NC_001477, NC_001474, NC_001475, NC_002640 해당 DNA 서열을 fasta 형식으로 <code>nuccore</code> 데이터베이스에서 다운로드 하시오. (참고로 <code>strwrap</code> 함수 사용법을 익혀두면 좋습니다)</p>
</div>
<div class="rmdnote">
<p><strong>Exercises </strong></p>
<ol style="list-style-type: decimal">
<li><p>popset 데이터베이스에서 “Covid-19” 단어가 들어간 유전자 40개를 찾고 (<code>entrez_search</code>에서 <code>retmax=40</code> 옵션 사용) 이들의 요약 정보 중 title 속성을 출력하시오 (<code>entrez_summary</code>와 <code>extract_from_esummary</code> 함수 사용).</p></li>
<li><p>위 결과에서 찾아진 유전자들 각각이 몇 개의 서열 샘플에 (population) 대해서 연구된 것인지 각각의 서열을 fasta 형태로 다운로드 받고 샘플의 개수에 대한 <code>barplot</code>을 그리시오</p></li>
</ol>
<ul>
<li>
<code>summary_record</code> 결과를 받아서 <code>extract_from_esummary</code>로 title을 추출 후 <code>data.frame</code>으로 변환</li>
<li>
<code>tidyverse</code>의 <code><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column()</a></code> 함수로 uid 정보 변수로 변환, mydata 이름으로 저장</li>
<li>
<code>entrez_fetch</code> 함수로 모든 uid에 대한 샘플 서열 <code>fasta</code> 파일 다운로드 후 파일 저장 (<code>write</code>함수 사용)</li>
<li>
<code>readDNAStringSet</code> 함수로 읽은 후 앞서 title 정보 비교를 통해서 앞서 mydata 와 병합</li>
<li>각 uid 별로 몇 개의 서열 샘플이 있는지 정보를 추출 후 barplot 그리기</li>
</ul>
</div>
<div class="rmdnote">
<p><strong>Exercises </strong></p>
<p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7938774/">Comparative sequence analysis of SARS-CoV-2 suggests its high transmissibility and pathogenicity</a> 논문을 참고하여 COVID-19 서열의 NCBI accession 번호를 찾고 <code>nuccore</code> 데이터베이스에서 <code>fasta</code> 포멧과 <code>genbank</code> 포멧의 정보를 다운로드 하시오. (데이터는 “covid_table.csv” 파일에 저장되어 있음)</p>
<div class="sourceCode" id="cb252"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">covid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="st">"Human"</span>, <span class="fl">7</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"Civet"</span>, <span class="st">"Civet"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">rep</a></span><span class="op">(</span><span class="st">"Bat"</span>, <span class="fl">3</span><span class="op">)</span>, <span class="st">"Pangolin"</span><span class="op">)</span>,
coronavirus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"SARS-CoV-2"</span>, <span class="st">"SARS-CoV-2"</span>, <span class="st">"SARS-CoV-1"</span>, <span class="st">"SARS-CoV-1"</span>, <span class="st">"SARS-CoV-1"</span>, <span class="st">"H-CoV-OC43"</span>, <span class="st">"MERS-CoV"</span>, <span class="st">"SARS-CoV"</span>, <span class="st">"SARS-CoV"</span>, <span class="st">"SL-CoV"</span>, <span class="st">"SL-CoV"</span>, <span class="st">"SL-CoV"</span>, <span class="st">"SL-CoV"</span><span class="op">)</span>,
isolate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"Wuhan Hu-1"</span>, <span class="st">"USA-WA-1"</span>, <span class="st">"Urbani"</span>, <span class="st">"Tor2"</span>, <span class="st">"GD03T10013"</span>, <span class="st">"UK/London"</span>,  <span class="st">"EMC-2012"</span>, <span class="st">"SZ3"</span>, <span class="st">"Civet007"</span>, <span class="st">"ZXC21"</span>, <span class="st">"WIV16"</span>, <span class="st">"RaTG13"</span>, <span class="st">"MP789"</span><span class="op">)</span>,
year <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"2020"</span>, <span class="st">"2020"</span>, <span class="st">"2002"</span>, <span class="st">"2002"</span>, <span class="st">"2003"</span>, <span class="st">"2011"</span>, <span class="st">"2011"</span>, <span class="st">"2003"</span>, <span class="st">"2004"</span>, <span class="st">"2015"</span>, <span class="st">"2013"</span>, <span class="st">"2013"</span>, <span class="st">"2020"</span><span class="op">)</span>,
gbacc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"NC_045512.2"</span>, <span class="st">"MN985325.1"</span>, <span class="st">"AY278741.1"</span>, <span class="st">"AY274119.3"</span>, <span class="st">"AY525636.1"</span>, <span class="st">"KU131570.1"</span>, <span class="st">"NC_019843.3"</span>, <span class="st">"AY304486.1"</span>, <span class="st">"AY572034.1"</span>, <span class="st">"MG772934.1"</span>, <span class="st">"KT444582.1"</span>, <span class="st">"MN996532.1"</span>, <span class="st">"MT084071.1"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.csv</a></span><span class="op">(</span><span class="va">covid</span>, file <span class="op">=</span> <span class="st">"covid_table.csv"</span>, quote <span class="op">=</span> <span class="cn">F</span>, row.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb253"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/">kableExtra</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: kableExtra</span>

<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/greendaygh/kribbr2022/main/covid_table.csv"</span>, destfile <span class="op">=</span> <span class="st">"covid_table2.csv"</span><span class="op">)</span>

<span class="va">covid19</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"covid_table2.csv"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/kableExtra/man/kable_classic.html">kable_classic</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">covid19</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="inline-table"><table class=" lightable-classic" style='font-family: "Arial Narrow", "Source Sans Pro", sans-serif; margin-left: auto; margin-right: auto;'>
<thead><tr>
<th style="text-align:left;">
species
</th>
<th style="text-align:left;">
coronavirus
</th>
<th style="text-align:left;">
isolate
</th>
<th style="text-align:right;">
year
</th>
<th style="text-align:left;">
gbacc
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
SARS-CoV-2
</td>
<td style="text-align:left;">
Wuhan Hu-1
</td>
<td style="text-align:right;">
2020
</td>
<td style="text-align:left;">
NC_045512.2
</td>
</tr>
<tr>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
SARS-CoV-2
</td>
<td style="text-align:left;">
USA-WA-1
</td>
<td style="text-align:right;">
2020
</td>
<td style="text-align:left;">
MN985325.1
</td>
</tr>
<tr>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
SARS-CoV-1
</td>
<td style="text-align:left;">
Urbani
</td>
<td style="text-align:right;">
2002
</td>
<td style="text-align:left;">
AY278741.1
</td>
</tr>
<tr>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
SARS-CoV-1
</td>
<td style="text-align:left;">
Tor2
</td>
<td style="text-align:right;">
2002
</td>
<td style="text-align:left;">
AY274119.3
</td>
</tr>
<tr>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
SARS-CoV-1
</td>
<td style="text-align:left;">
GD03T10013
</td>
<td style="text-align:right;">
2003
</td>
<td style="text-align:left;">
AY525636.1
</td>
</tr>
<tr>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
H-CoV-OC43
</td>
<td style="text-align:left;">
UK/London
</td>
<td style="text-align:right;">
2011
</td>
<td style="text-align:left;">
KU131570.1
</td>
</tr>
<tr>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
MERS-CoV
</td>
<td style="text-align:left;">
EMC-2012
</td>
<td style="text-align:right;">
2011
</td>
<td style="text-align:left;">
NC_019843.3
</td>
</tr>
<tr>
<td style="text-align:left;">
Civet
</td>
<td style="text-align:left;">
SARS-CoV
</td>
<td style="text-align:left;">
SZ3
</td>
<td style="text-align:right;">
2003
</td>
<td style="text-align:left;">
AY304486.1
</td>
</tr>
<tr>
<td style="text-align:left;">
Civet
</td>
<td style="text-align:left;">
SARS-CoV
</td>
<td style="text-align:left;">
Civet007
</td>
<td style="text-align:right;">
2004
</td>
<td style="text-align:left;">
AY572034.1
</td>
</tr>
<tr>
<td style="text-align:left;">
Bat
</td>
<td style="text-align:left;">
SL-CoV
</td>
<td style="text-align:left;">
ZXC21
</td>
<td style="text-align:right;">
2015
</td>
<td style="text-align:left;">
MG772934.1
</td>
</tr>
<tr>
<td style="text-align:left;">
Bat
</td>
<td style="text-align:left;">
SL-CoV
</td>
<td style="text-align:left;">
WIV16
</td>
<td style="text-align:right;">
2013
</td>
<td style="text-align:left;">
KT444582.1
</td>
</tr>
<tr>
<td style="text-align:left;">
Bat
</td>
<td style="text-align:left;">
SL-CoV
</td>
<td style="text-align:left;">
RaTG13
</td>
<td style="text-align:right;">
2013
</td>
<td style="text-align:left;">
MN996532.1
</td>
</tr>
<tr>
<td style="text-align:left;">
Pangolin
</td>
<td style="text-align:left;">
SL-CoV
</td>
<td style="text-align:left;">
MP789
</td>
<td style="text-align:right;">
2020
</td>
<td style="text-align:left;">
MT084071.1
</td>
</tr>
</tbody>
</table></div>
</div>
</div>
<div id="align-two-sequences" class="section level2" number="12.2">
<h2>
<span class="header-section-number">12.2</span> Align two sequences<a class="anchor" aria-label="anchor" href="#align-two-sequences"><i class="fas fa-link"></i></a>
</h2>
<p>서열 정렬은 match, mismatch, penalty 등의 scoring rule을 기반으로 최적의 score를 갖는 서열 정렬을 찾는 알고리즘 입니다. Biostrings 패키지에는 두 개의 서열에 대해 local, global alignment를 수행할 수 있는 <code>pairwiseAlignment</code> 함수를 제공하고 있습니다. 첫 번째 파라메터는 pattern이며 두 번째는 subject 로서 pattern은 query로서 해당 서열이 subject (target)에 있는지를 보는 것과 같습니다.</p>
<div class="sourceCode" id="cb254"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">covid19seq</span>

<span class="op">?</span><span class="va">pairwiseAlignment</span>
<span class="va">aln</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/pairwiseAlignment.html">pairwiseAlignment</a></span><span class="op">(</span><span class="va">covid19seq</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">covid19seq</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/methods.html">methods</a></span><span class="op">(</span>class<span class="op">=</span><span class="st">"PairwiseAlignmentsSingleSubject"</span><span class="op">)</span>
<span class="op">?</span><span class="va">PairwiseAlignmentsSingleSubject</span></code></pre></div>
<p>위에서 서열 정렬이 된 결과를 <code>DNAString</code> class의 변수에 저장한 후 해당 class에서 제공하는 다양한 함수를 동일하게 적용할 수 있습니다. 또한 <code>writePairwiseAlignments</code> 함수는 두 서열의 비교 결과를 보기 좋게 출력해주는 기능을 수행합니다. <code>summary</code> 함수를 사용하면 염기가 다른 곳의 위치를 출력해주며 <code>consensusString</code>은 50% 초과하는 서열에 대한 문자열을 출력해 줍니다. 이 외에도 <code>score</code>, <code>consensusMatrix</code> 등 다양한 help 함수들이 있습니다.</p>
<div class="sourceCode" id="cb255"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alnseqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/PairwiseAlignments-class.html">alignedPattern</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/PairwiseAlignments-class.html">alignedSubject</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">alnseqs</span><span class="op">)</span>

<span class="va">aln</span>

<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/PairwiseAlignments-io.html">writePairwiseAlignments</a></span><span class="op">(</span><span class="va">aln</span>, block.width<span class="op">=</span><span class="fl">50</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/PairwiseAlignments-io.html">writePairwiseAlignments</a></span><span class="op">(</span><span class="va">aln</span>, file<span class="op">=</span><span class="st">"covidalign.txt"</span>,  block.width<span class="op">=</span><span class="fl">30</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html">consensusMatrix</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html">consensusString</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/score.html">score</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span></code></pre></div>
<p>참고로 아래와 같이 별도의 정보 테이블을 만들고 서열의 이름은 간단히 id 만 사용해서 분석하는 것이 더 효율적입니다. 문자열을 다루는 코드를 익혀두시기 바랍니다.</p>
<div class="sourceCode" id="cb256"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">covid19seq</span><span class="op">)</span>

<span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">covid19seq</span><span class="op">)</span>, split<span class="op">=</span><span class="st">" "</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">}</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> 
  <span class="va">unlist</span>

<span class="va">titles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">covid19seq</span><span class="op">)</span>, split<span class="op">=</span><span class="st">" "</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">lapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>, collapse<span class="op">=</span><span class="st">" "</span><span class="op">)</span>
    <span class="op">}</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> 
  <span class="va">unlist</span>

<span class="va">covid19info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">ids</span>, <span class="va">titles</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">covid19seq</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">covid19info</span><span class="op">$</span><span class="va">ids</span>

<span class="va">aln</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/pairwiseAlignment.html">pairwiseAlignment</a></span><span class="op">(</span><span class="va">covid19seq</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">covid19seq</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/PairwiseAlignments-io.html">writePairwiseAlignments</a></span><span class="op">(</span><span class="va">aln</span>, block.width<span class="op">=</span><span class="fl">50</span><span class="op">)</span></code></pre></div>
</div>
<div id="multiple-sequence-alignment" class="section level2" number="12.3">
<h2>
<span class="header-section-number">12.3</span> Multiple sequence alignment<a class="anchor" aria-label="anchor" href="#multiple-sequence-alignment"><i class="fas fa-link"></i></a>
</h2>
<p>Multiple sequence alignment(MSA) tool은 서열 데이터의 양과 계산량의 문제로 linux 기반 commandline 프로그램들이 많습니다. 대표적으로 <a href="https://www.ebi.ac.uk/Tools/msa/clustalo/">CLUSTAL-Omega</a>, <a href="https://www.ebi.ac.uk/Tools/msa/muscle/">MUSCLE</a>. window 기반 환경에서는 docker 등을 활용해서 관련 분석을 수행할 수 있습니다.</p>
<div id="msa" class="section level3" number="12.3.1">
<h3>
<span class="header-section-number">12.3.1</span> msa<a class="anchor" aria-label="anchor" href="#msa"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/msa.html">msa</a> 패키지는 위 Clustal나 MUSCLE 등의 프로그램에 대한 R 인터페이스를 제공하며 Linux 뿐만 아니라 모든 운영체제에서 수행될 수 있도록 만들어 둔 패키지 입니다.</p>
<div class="sourceCode" id="cb257"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.bioinf.jku.at/software/msa/">msa</a></span><span class="op">)</span>

<span class="va">alnmsa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msa/man/msa.html">msa</a></span><span class="op">(</span><span class="va">subcovid19seq</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">alnmsa</span><span class="op">)</span>
<span class="va">alnmsa</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">alnmsa</span>, show<span class="op">=</span><span class="st">"complete"</span><span class="op">)</span>
<span class="op">?</span><span class="va">msaPrettyPrint</span>
<span class="fu"><a href="https://rdrr.io/pkg/msa/man/msaPrettyPrint.html">msaPrettyPrint</a></span><span class="op">(</span><span class="va">alnmsa</span>, output<span class="op">=</span><span class="st">"pdf"</span>, showNames<span class="op">=</span><span class="st">"none"</span>, showLogo<span class="op">=</span><span class="st">"top"</span>, askForOverwrite<span class="op">=</span><span class="cn">FALSE</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">myconseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msa/man/msaConsensusSequence-methods.html">msaConsensusSequence</a></span><span class="op">(</span><span class="va">alnmsa</span><span class="op">)</span></code></pre></div>
<p>참고로 정렬된 서열 출력물에 표시되는 ?는 판단하기 모호한 위치를 나타냅니다. 예를 들어 A와 T 가 5:5로 나타난 위치는 ?로 표시됩니다.</p>
<p><code>MsaDNAMultipleAlignment</code> class는 Biostring 패키지의 <code>DNAString</code> class를 상속받은 클래스로서 다음과 같이 <code>DNAStringSet</code> class로 변환해서 분석도 가능합니다.</p>
<div class="sourceCode" id="cb258"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alnseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">alnseq</span><span class="op">)</span>
<span class="va">alnseq</span>
<span class="va">myconseq2</span> <span class="op">&lt;-</span> <span class="fu">ConsensusSequence</span><span class="op">(</span><span class="va">alnseq</span><span class="op">)</span></code></pre></div>
<p>위 alignment 결과에서 관심있는 특정 위치만을 선택해서 임의의 분석을 수행하고 싶은 경우 마스킹 함수를 사용할 수 있습니다. 이 기능 역시 <code>MsaDNAMultipleAlignment</code> class는 <code>Biostring</code> 패키지의 <code>DNAString</code> class 모두에 적용이 가능합니다. IRanges 함수는 뒤에서 더 상세히 설명하도록 하겠습니다.</p>
<div class="sourceCode" id="cb259"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">colM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges</a></span><span class="op">(</span>start<span class="op">=</span><span class="fl">1</span>, end<span class="op">=</span><span class="fl">300</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/MultipleAlignment-class.html">colmask</a></span><span class="op">(</span><span class="va">alnmsa</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">colM</span>
<span class="va">alnmsa</span>
<span class="fu"><a href="https://rdrr.io/pkg/msa/man/msaConsensusSequence-methods.html">msaConsensusSequence</a></span><span class="op">(</span><span class="va">alnmsa</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html">alphabetFrequency</a></span><span class="op">(</span><span class="va">alnmsa</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html">alphabetFrequency</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/MaskedXString-class.html">unmasked</a></span><span class="op">(</span><span class="va">alnmsa</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="decipher" class="section level3" number="12.3.2">
<h3>
<span class="header-section-number">12.3.2</span> DECIPHER<a class="anchor" aria-label="anchor" href="#decipher"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://www.bioconductor.org/packages/release/bioc/html/DECIPHER.html">DECIPHER</a> 패키지는 서열 alignment나 primer design 등을 수행할 수 있는 패키지로 다음과 같이 별도 메모리에 서열을 저장하고 빠르게 alignment를 수행할 수 있어서 중소 규모의 서열에 대한 분석으로 유용하게 사용될 수 있습니다. 다음은 관련 서열을 SQLite 데이터베이스에 저장하고 그 내용을 쉽게 볼 수 있는 기능들 입니다. <code>dbDisconnect</code> 함수를 실행하면 모든 저장된 데이터가 사라지며 매모리는 다시 사용할 수 있게 됩니다.</p>
<div class="sourceCode" id="cb260"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">DECIPHER</span><span class="op">)</span>

<span class="va">dbConn</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">":memory:"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">covid19seq</span>, <span class="st">"XStringSet"</span>, <span class="va">dbConn</span>, <span class="st">"covid19"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span>

<span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/IdLengths.html">IdLengths</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/Add2DB.html">Add2DB</a></span><span class="op">(</span><span class="va">l</span>, <span class="va">dbConn</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">trnl</span>, <span class="st">"XStringSet"</span>, <span class="va">dbConn</span>, identifier <span class="op">=</span> <span class="st">"trnl"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/Seqs2DB.html">Seqs2DB</a></span><span class="op">(</span><span class="va">coi</span>, <span class="st">"XStringSet"</span>, <span class="va">dbConn</span>, <span class="st">"coi"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span>

<span class="va">l</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/IdLengths.html">IdLengths</a></span><span class="op">(</span><span class="va">dbConn</span>, identifier <span class="op">=</span> <span class="st">"trnl"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/Add2DB.html">Add2DB</a></span><span class="op">(</span><span class="va">l</span>, <span class="va">dbConn</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseDB.html">BrowseDB</a></span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span>

<span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">dbConn</span><span class="op">)</span></code></pre></div>
<p>다중서열을 정렬하는 <code>AlignSeqs</code> 함수와 <code>BrowseSeqs</code> 함수를 활용해서 html 형태로 정렬된 서열을 볼 수 있습니다. 특히 <code>patterns</code> 옵션을 사용해서 원하는 서열이 존재하는지도 확인할 수 있습니다.</p>
<div class="sourceCode" id="cb261"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">## extract sequences</span>
<span class="va">covid19seq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/SearchDB.html">SearchDB</a></span><span class="op">(</span><span class="va">dbConn</span>, identifier <span class="op">=</span> <span class="st">"covid19"</span><span class="op">)</span>
<span class="va">subcovid19seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/XVector/man/XVector-class.html">subseq</a></span><span class="op">(</span><span class="va">covid19seq2</span>, <span class="fl">1</span>, <span class="fl">2000</span><span class="op">)</span>
<span class="va">aln</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/AlignSeqs.html">AlignSeqs</a></span><span class="op">(</span><span class="va">subcovid19seq</span><span class="op">)</span>
<span class="va">aln</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">aln</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">aln</span>, colWidth <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">aln</span>, colWidth <span class="op">=</span> <span class="fl">100</span>, patterns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"ACTG"</span>, <span class="st">"CSC"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">aln</span>, colWidth <span class="op">=</span> <span class="fl">100</span>, patterns<span class="op">=</span><span class="st">"-"</span>, colors<span class="op">=</span><span class="st">"black"</span><span class="op">)</span>
<span class="op">?</span><span class="va">BrowseSeqs</span>

<span class="va">aln2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/AlignTranslation.html">AlignTranslation</a></span><span class="op">(</span><span class="va">subcovid19seq</span>, type<span class="op">=</span><span class="st">"AAStringSet"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">aln2</span>, colWidth <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">aln2</span>, colWidth <span class="op">=</span> <span class="fl">100</span>, highlight<span class="op">=</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><code>AlignSeqs</code>함수의 결과가 <code>DNAStringSet</code> 클래스이므로 앞서 수행한 마스킹 등의 기능을 동일하게 적용 가능합니다.</p>
<p><code>DECIPHER</code> 패키지의 <code>DigestDNA</code>함수를 이용하면 enzyme digestion을 시뮬레이션할 수 있는 기능을 활용할 수 있습니다. 단 숫자 등 필요없는 문자를 제거하기 위해서 stringr 패키지를 사용합니다.</p>
<div class="sourceCode" id="cb262"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">RESTRICTION_ENZYMES</span><span class="op">)</span>
<span class="va">RESTRICTION_ENZYMES</span>
<span class="va">rsite</span> <span class="op">&lt;-</span> <span class="va">RESTRICTION_ENZYMES</span><span class="op">[</span><span class="st">"BsmBI"</span><span class="op">]</span>
<span class="va">rsite</span> <span class="op">&lt;-</span> <span class="va">RESTRICTION_ENZYMES</span><span class="op">[</span><span class="st">"BsaI"</span><span class="op">]</span>

<span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/DigestDNA.html">DigestDNA</a></span><span class="op">(</span><span class="va">rsite</span>, <span class="va">covid19seq2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html">unlist</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span>
<span class="co">#writeXStringSet(unlist(d), file="covid19bsmbi.fasta")</span>
<span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/DigestDNA.html">DigestDNA</a></span><span class="op">(</span><span class="va">rsite</span>, <span class="va">covid19seq2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, type<span class="op">=</span><span class="st">"positions"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html">unlist</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span>


<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://stringr.tidyverse.org">stringr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringi.gagolewski.com/">stringi</a></span><span class="op">)</span>


<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">covid19seq2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, colWidth <span class="op">=</span> <span class="fl">100</span>, patterns<span class="op">=</span><span class="va">rsite</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span><span class="op">(</span><span class="st">"(^[[:alpha:]]*).*"</span>, <span class="st">""</span>, <span class="va">rsite</span><span class="op">)</span>

<span class="va">rsite2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html">str_extract_all</a></span><span class="op">(</span><span class="va">rsite</span>, <span class="st">"[[A-Z]]"</span>, simplify <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, collapse<span class="op">=</span><span class="st">""</span><span class="op">)</span>
<span class="va">rsite3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/reverseComplement.html">reverseComplement</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html">DNAString</a></span><span class="op">(</span><span class="va">rsite2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/BrowseSeqs.html">BrowseSeqs</a></span><span class="op">(</span><span class="va">covid19seq2</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, colWidth <span class="op">=</span> <span class="fl">100</span>, patterns<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="va">rsite2</span>, <span class="va">rsite3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="rmdnote">
<p><strong>Exercises </strong></p>
<ol style="list-style-type: decimal">
<li><p>앞서 다운로드 받은 Latrodectus katipo 서열 데이터를 읽어들이고 100bp 단위로 출력하시오</p></li>
<li><p>MSA 를 수행하고 정렬된 결과를 100bp 단위로 출력하시오</p></li>
<li><p><code>ConsensusSequence</code> 함수를 이용하여 정렬된 결과로부터 consensus 서열을 추출하시오</p></li>
</ol>
</div>
<p>참고로 염기와 아미노산의 Standard Ambiguity Codes는 각각 다음과 같습니다.</p>
<ul>
<li><div class="inline-figure">DNA ambiguity code
<img src="images/dnasscode.png">
</div></li>
<li><div class="inline-figure">Amino acid ambiguity code
<img src="images/aaacode.png">
</div></li>
</ul>
</div>
</div>
<div id="phylogenetic-trees-with-clustering" class="section level2" number="12.4">
<h2>
<span class="header-section-number">12.4</span> Phylogenetic trees with clustering<a class="anchor" aria-label="anchor" href="#phylogenetic-trees-with-clustering"><i class="fas fa-link"></i></a>
</h2>
<p>다중서열비교 결과는 계통학에서 널리 쓰이며 <code>msa</code>나 <code>DECIPHER</code> 패키지에서 얻어진 결과를 계통학의 tree 형태로 가시화할 수 있습니다. tree 형태의 가시화를 위해 다양한 포멧의 파일이 개발되었고 <code>treeio</code> 패키지는 이들 다양한 포맷의 파일을 쉽게 변환하기 위해 만들어진 패키지 입니다. 다음은 Newick tree 포멧의 예 입니다.</p>
<pre><code>    ((t2:0.04,t1:0.34):0.89,(t5:0.37,(t4:0.03,t3:0.67):0.9):0.59); </code></pre>
<p>가장 널리 사용되는 포멧은 <code>phylo</code> 형태로서 이는 <code>ape</code>라는 phylogenetic 분석의 대표적인 패키지에서 처음 제안되어 사용되고 있습니다. 최근 ggplot 형태의 <a href="https://yulab-smu.top/treedata-book/chapter12.html">ggtree</a>, <a href="https://www.molecularecologist.com/2017/02/08/phylogenetic-trees-in-r-using-ggtree/">reference</a>이 개발되어 계통도를 좀더 세밀하게 그릴 수 있으며 ggtree는 <code>phylo</code> 형태의 포맷을 주로 사용합니다.</p>
<div class="sourceCode" id="cb264"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ggtree"</span><span class="op">)</span>
<span class="co">#BiocManager::install("treeio")</span></code></pre></div>
<p>Phylogenetic tree는 서열간의 유사도(거리)를 기반으로 분석할 수 있습니다. 앞서 사용한 <code>mas</code>나 <code>DECIPHER</code> 패키지로 얻어진 MSA 결과는 모두 <code>Biostrings</code> 패키지의 <code>XStringSet</code> (<code>DNAStringSet</code>, <code>AAStringSet</code> 등) 클래스 입니다. 따라서 XStringSet의 거리를 계산해주는 <code><a href="https://rdrr.io/pkg/Biostrings/man/stringDist.html">Biostrings::stringDist</a></code> 함수나 <code><a href="https://rdrr.io/pkg/DECIPHER/man/DistanceMatrix.html">DECIPHER::DistanceMatrix</a></code>가 사용될 수 있습니다. 참고로 <code>phylo</code> class는 <code>as.tibble</code> 함수를 이용해서 테이블 형태로 변환, 활용할 수 있습니다.</p>
<div class="sourceCode" id="cb265"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/">ape</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/treedata-book/">ggtree</a></span><span class="op">)</span>

<span class="va">alnmsa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msa/man/msa.html">msa</a></span><span class="op">(</span><span class="va">subcovid19seq</span><span class="op">)</span>
<span class="va">mydist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/stringDist.html">stringDist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span><span class="op">(</span><span class="va">alnmsa</span><span class="op">)</span><span class="op">)</span>
<span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">mydist</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span>

<span class="va">mytree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.phylo.html">as.phylo</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">mytree</span>, layout<span class="op">=</span><span class="st">"circular"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="https://tibble.tidyverse.org/reference/deprecated.html">as.tibble</a></span><span class="op">(</span><span class="va">mytree</span><span class="op">)</span></code></pre></div>
<p><code>DECIPHER</code> 패키지에는 <code>XStringSet</code> 서열의 거리를 계산해주는 <code>DistanceMatrix</code> 함수가 있습니다. 이 함수를 이용하면 역시 같은 패키지에서 제공하는 <code>IdClusters</code> 함수를 이용해서 유사한 서열끼리 묶어주는 tree 를 만들 수 있습니다. <code>dendrogram</code>는 <code>str</code> 함수 활용이 가능합니다.</p>
<div class="sourceCode" id="cb266"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/DistanceMatrix.html">DistanceMatrix</a></span><span class="op">(</span><span class="va">subcovid19seq</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span>

<span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DECIPHER/man/IdClusters.html">IdClusters</a></span><span class="op">(</span><span class="va">dm</span>, cutoff<span class="op">=</span><span class="fl">10</span>, method<span class="op">=</span><span class="st">"NJ"</span>, showPlot<span class="op">=</span><span class="cn">F</span>, type<span class="op">=</span><span class="st">"dendrogram"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/methods.html">methods</a></span><span class="op">(</span>class<span class="op">=</span><span class="st">"dendrogram"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span></code></pre></div>
<p><code>dendrogram</code> class는 <code>hclust</code>를 거처 <code>phylo</code> class 형태로 변환 후 <code>ggtree</code> 패키지를 활용할 수 있습니다.</p>
<div class="sourceCode" id="cb267"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## convert to dendrogram -&gt; hclust -&gt; phylo </span>
<span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/as.hclust.html">as.hclust</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span>
<span class="va">py</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.phylo.html">as.phylo</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">py</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">py</span><span class="op">)</span>
<span class="fu"><a href="https://tibble.tidyverse.org/reference/deprecated.html">as.tibble</a></span><span class="op">(</span><span class="va">py</span><span class="op">)</span></code></pre></div>
<p>ggtree를 활용하면 다양한 레이아웃을 활용할 수 있고 레이아웃에 대한 정보는 <a href="https://yulab-smu.top/treedata-book/chapter4.html#tree-layouts">Layouts of a phylogenetic tree</a> 이 곳을 참고하시면 되겠습니다.</p>
<div class="sourceCode" id="cb268"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/rtree.html">rtree</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tree.html">geom_tree</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/theme_tree.html">theme_tree</a></span><span class="op">(</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span>, branch.length<span class="op">=</span><span class="st">"none"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"firebrick"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>angle<span class="op">=</span><span class="va">angle</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span>, branch.length<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>angle<span class="op">=</span><span class="va">angle</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/theme_tree2.html">theme_tree2</a></span><span class="op">(</span><span class="op">)</span> 

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span>, layout<span class="op">=</span><span class="st">"circular"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">100</span>,<span class="fl">30</span>,<span class="fl">100</span>,<span class="fl">30</span><span class="op">)</span>, <span class="st">"mm"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span><span class="op">(</span><span class="st">"myphylo.pdf"</span>, width <span class="op">=</span> <span class="fl">50</span>, height <span class="op">=</span> <span class="fl">50</span>, units <span class="op">=</span> <span class="st">"cm"</span>, limitsize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>covid19 example</p>
<div class="sourceCode" id="cb269"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">covid19seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="st">"covid19.fasta"</span>, format<span class="op">=</span><span class="st">"fasta"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>
<span class="va">subcovid19seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/XVector/man/XVector-class.html">subseq</a></span><span class="op">(</span><span class="va">covid19seq</span>, <span class="fl">1</span>, <span class="fl">2000</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">subcovid19seq</span><span class="op">)</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">subcovid19seq</span><span class="op">)</span>, <span class="st">" "</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">}</span><span class="op">)</span>
  
<span class="va">alnmsa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msa/man/msa.html">msa</a></span><span class="op">(</span><span class="va">subcovid19seq</span><span class="op">)</span>
<span class="va">mydist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/stringDist.html">stringDist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html">DNAStringSet</a></span><span class="op">(</span><span class="va">alnmsa</span><span class="op">)</span><span class="op">)</span>
<span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html">hclust</a></span><span class="op">(</span><span class="va">mydist</span><span class="op">)</span>
<span class="va">mytree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/as.phylo.html">as.phylo</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">mytree</span>, layout<span class="op">=</span><span class="st">"circular"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"firebrick"</span>, size<span class="op">=</span><span class="fl">3</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">mytree</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span>color<span class="op">=</span><span class="st">"firebrick"</span>, size<span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/theme_tree2.html">theme_tree2</a></span><span class="op">(</span>scale<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span> </code></pre></div>
<p>특정 그룹을 highlight 하기 위해서 <code>geom_hilight</code> 함수를 사용합니다.</p>
<div class="sourceCode" id="cb270"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/theme_tree2.html">theme_tree2</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom-hilight.html">geom_hilight</a></span><span class="op">(</span>node<span class="op">=</span><span class="fl">20</span>, fill<span class="op">=</span><span class="st">"steelblue"</span>, alpha<span class="op">=</span><span class="fl">.4</span><span class="op">)</span> </code></pre></div>
<p>노드를 알아보기 위해서 <code>tidytree</code>를 사용할 수 있습니다.</p>
<div class="sourceCode" id="cb271"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://tibble.tidyverse.org/reference/deprecated.html">as.tibble</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>

<span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>node<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">20</span>, <span class="fl">22</span><span class="op">)</span>, type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"T1"</span>, <span class="st">"T1"</span>, <span class="st">"T2"</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/theme_tree2.html">theme_tree2</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom-hilight.html">geom_hilight</a></span><span class="op">(</span><span class="va">d</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>node<span class="op">=</span><span class="va">node</span>, fill<span class="op">=</span><span class="va">type</span><span class="op">)</span>, alpha<span class="op">=</span><span class="fl">.4</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="st">"steelblue"</span>, <span class="st">"darkgreen"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div id="facet-utilities" class="section level3" number="12.4.1">
<h3>
<span class="header-section-number">12.4.1</span> Facet Utilities<a class="anchor" aria-label="anchor" href="#facet-utilities"><i class="fas fa-link"></i></a>
</h3>
<p>geom_facet 와 facet_widths 를 사용하면 추가 판넬에 tree를 그릴 수 있습니다.</p>
<div class="sourceCode" id="cb272"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/rtree.html">rtree</a></span><span class="op">(</span><span class="fl">30</span><span class="op">)</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html">ggtree</a></span><span class="op">(</span><span class="va">tree</span>, branch.length <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html">geom_tiplab</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">'none'</span><span class="op">)</span>

<span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>
<span class="va">b</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">a</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span>, <span class="va">a</span>, <span class="va">b</span><span class="op">)</span>
<span class="va">df2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span><span class="va">df</span>, <span class="op">-</span><span class="va">tree.tip.label</span><span class="op">)</span>

<span class="va">p2</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/facet-plot.html">geom_facet</a></span><span class="op">(</span>panel <span class="op">=</span> <span class="st">'bar'</span>, data <span class="op">=</span> <span class="va">df2</span>, geom <span class="op">=</span> <span class="va">geom_bar</span>, 
                 mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">name</span><span class="op">)</span><span class="op">)</span>, 
                 orientation <span class="op">=</span> <span class="st">'y'</span>, width <span class="op">=</span> <span class="fl">0.8</span>, stat<span class="op">=</span><span class="st">'identity'</span><span class="op">)</span> <span class="op">+</span> 
        <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/xlim_tree.html">xlim_tree</a></span><span class="op">(</span><span class="fl">9</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/facet_widths.html">facet_widths</a></span><span class="op">(</span><span class="va">p2</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="blast-result-analysis" class="section level2" number="12.5">
<h2>
<span class="header-section-number">12.5</span> BLAST result analysis<a class="anchor" aria-label="anchor" href="#blast-result-analysis"><i class="fas fa-link"></i></a>
</h2>
<p>BLAST를 로컬컴퓨터에 설치하거나 docker를 이용해서 활용할 수 있으나 본 강의에서는 직접 BLAST를 수행하는 대신 NCBI에서 실행한 BLAST 출력물을 분석하는 경우에 대해서 설명을 진행하겠습니다. 예시로는 PET를 분해하는 단백질로 알려진 IsPETase의 서열과 유사한 서열을 찾아서 분석해 보겠습니다. IsPETase 정보는 다음과 같습니다 Genes encoding I. sakaiensis 201-F6 IsPETase (WT PETase) (accession number: A0A0K8P6T7).</p>
<div class="rmdnote">
<p><strong>Exercises </strong></p>
<ol style="list-style-type: decimal">
<li><p>NCBI BLAST 사이트에서 A0A0K8P6T7 단백질에 대한 BLASTp를 수행하시오 (db: nr)</p></li>
<li><p>결과물을 (100개) 다운로드 하고 (fasta와 hit table 각 1개 파일씩) 작업디렉토리로 복사하시오</p></li>
<li><p>fasta 와 hit 데이터를 각각 읽어들이시오</p></li>
<li><p>100개의 서열을 DECIPHER 패키지를 활용해서 정렬하고 100bp 단위로 출력해보시오</p></li>
<li><p>align된 결과에서 consensus 서열을 추출하고 각 위치별로 어떤 아미노산이 많은지 bar 그래프를 그려보시오</p></li>
<li><p>align된 결과를 <code>ggtree</code> 패키지를 사용해서 phylogenetic를 그리시오</p></li>
</ol>
</div>
<hr>
<p><a rel="license" href="http://creativecommons.org/licenses/by-nc-nd/4.0/"><img alt="크리에이티브 커먼즈 라이선스" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-nd/4.0/88x31.png"></a><br>이 저작물은 <a rel="license" href="http://creativecommons.org/licenses/by-nc-nd/4.0/">크리에이티브 커먼즈 저작자표시-비영리-변경금지 4.0 국제 라이선스</a>에 따라 이용할 수 있습니다.</p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="biostrings.html"><span class="header-section-number">11</span> Biostrings</a></div>
<div class="next"><a href="tools-for-genome-analysis.html"><span class="header-section-number">13</span> Tools for genome analysis</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#tools-for-sequences"><span class="header-section-number">12</span> Tools for sequences</a></li>
<li><a class="nav-link" href="#sequences-from-ncbi"><span class="header-section-number">12.1</span> Sequences from NCBI</a></li>
<li><a class="nav-link" href="#align-two-sequences"><span class="header-section-number">12.2</span> Align two sequences</a></li>
<li>
<a class="nav-link" href="#multiple-sequence-alignment"><span class="header-section-number">12.3</span> Multiple sequence alignment</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#msa"><span class="header-section-number">12.3.1</span> msa</a></li>
<li><a class="nav-link" href="#decipher"><span class="header-section-number">12.3.2</span> DECIPHER</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#phylogenetic-trees-with-clustering"><span class="header-section-number">12.4</span> Phylogenetic trees with clustering</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#facet-utilities"><span class="header-section-number">12.4.1</span> Facet Utilities</a></li></ul>
</li>
<li><a class="nav-link" href="#blast-result-analysis"><span class="header-section-number">12.5</span> BLAST result analysis</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>2022년 한국생명공학연구원 연구데이터 분석과정 R</strong>" was written by 합성생물학전문연구소 김하성. It was last built on 2022-10-20.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
